src/coreclr/jit/morph.cpp

```diff
@@ -4691,12 +4879,65 @@ GenTree* Compiler::fgMorphMultiregStructArg(GenTree* arg, fgArgTabEntry* fgEntry
             //
             unsigned offset = baseOffset;
             newArg          = new (this, GT_FIELD_LIST) GenTreeFieldList();
+#if defined(TARGET_LOONGARCH64)
+            uint32_t floatFieldFlags = info.compCompHnd->getLoongArch64PassStructInRegisterFlags(objClass);
+            if (floatFieldFlags & (STRUCT_HAS_FLOAT_FIELDS_MASK ^ STRUCT_FLOAT_FIELD_ONLY_ONE))
+            {
+                assert((floatFieldFlags & STRUCT_HAS_FLOAT_FIELDS_MASK) > 1);
+                var_types tmp_type_1;
+                var_types tmp_type_2;
+
+                compFloatingPointUsed = true;
+                if (floatFieldFlags & STRUCT_FLOAT_FIELD_ONLY_TWO)
+                {
+                    tmp_type_1 = (floatFieldFlags & STRUCT_FIRST_FIELD_SIZE_IS8) ? TYP_DOUBLE : TYP_FLOAT;
+                    tmp_type_2 = (floatFieldFlags & STRUCT_SECOND_FIELD_SIZE_IS8) ? TYP_DOUBLE : TYP_FLOAT;
+                }
+                else if (floatFieldFlags & STRUCT_FLOAT_FIELD_FIRST)
+                {
+                    tmp_type_1 = (floatFieldFlags & STRUCT_FIRST_FIELD_SIZE_IS8) ? TYP_DOUBLE : TYP_FLOAT;
+                    tmp_type_2 = (floatFieldFlags & STRUCT_SECOND_FIELD_SIZE_IS8) ? type[1] : TYP_INT;
+                }
+                else if (floatFieldFlags & STRUCT_FLOAT_FIELD_SECOND)
+                {
+                    tmp_type_1 = (floatFieldFlags & STRUCT_FIRST_FIELD_SIZE_IS8) ? type[0] : TYP_INT;
+                    tmp_type_2 = (floatFieldFlags & STRUCT_SECOND_FIELD_SIZE_IS8) ? TYP_DOUBLE : TYP_FLOAT;
+                }
+                else
+                {
+                    NYI_LOONGARCH64("fgMorphMultiregStructArg -- GT_LCL_FLD,GT_LCL_VAR");
+                    tmp_type_1 = TYP_UNDEF;
+                    tmp_type_2 = TYP_UNDEF;
+                }
+                elemSize = (floatFieldFlags & STRUCT_HAS_8BYTES_FIELDS_MASK) ? 8 : 4;
+
+                GenTree* nextLclFld = gtNewLclFldNode(varNum, tmp_type_1, offset);
+                newArg->AddField(this, nextLclFld, offset, tmp_type_1);
+                offset += elemSize;
+                nextLclFld = gtNewLclFldNode(varNum, tmp_type_2, offset);
+                newArg->AddField(this, nextLclFld, offset, tmp_type_2);
+            }
+            else
+            {
+                GenTree* nextLclFld = gtNewLclFldNode(varNum, type[0], offset);
+                newArg->AddField(this, nextLclFld, offset, type[0]);
+
+                if (elemCount > 1)
+                {
+                    assert(elemCount == 2);
+                    elemSize   = genTypeSize(type[1]);
+                    nextLclFld = gtNewLclFldNode(varNum, type[1], offset + elemSize);
+                    newArg->AddField(this, nextLclFld, offset + elemSize, type[1]);
+                }
+            }
+#else
```

Comment:
This is not the right place for this code.

The idea with this method is that the ABI-specific code "fills in" the `type[inx]` array, and then the loops below "unroll" that into the `FIELD_LIST`. See e. g. how Unix x64 handles this.

LA should follow the same pattern.

Reply:
I know that there is an improvement for LA-ABI.
Maybe it's better do it later by a new patch.

Reply:
I had deleted these.

Thanks.